cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -353,22 +385,49 @@ public void apply() throws Exception {
    * @throws Exception
    */
   public void upgrade()
-    throws InterruptedException, TransactionFailureException, IOException, DatasetManagementException {
-    initialize();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() {
-          upgradeVersionKeys();
-        }
-      });
+    throws InterruptedException, IOException, DatasetManagementException {
+    LOG.info("Starting upgrade of DatasetBasedStreamSizeScheduleStore.");
+
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.MINUTES.sleep(1);
+    }
+
+    while (!appVerUpgrade.get()) {
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
+          .execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() {
+              if (upgradeVersionKeys(table)) {
+                // Upgrade is complete. Mark that app version upgrade is complete in the table.
+                table.put(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'), APP_VERSION_UPGRADE_KEY);
+              }
+            }
+          });
+      } catch (TransactionFailureException e) {
+        LOG.debug("Upgrade step faced Transaction failure. Will retry operation after some delay.", e);
```

Comment:
there are various possibilities:
- conflict: in this case you want to retry right away. 
- timeout: you want to try again with longer timeout? Or with fewer records?
- other errors: thse would be unrelated to transactions, and I think not retryable. They should be logged as errors not debug.

Also, can other operations fail with conflicts while the upgrade is running)?